apis/goharbor.io/v1alpha2/harbor_types.go

```diff
@@ -0,0 +1,461 @@
+			Database:            r.Prefix + databaseName,
+			Hosts:               r.Hosts,
+		},
+		Parameters: map[string]string{
+			harbormetav1.PostgresSSLModeKey: string(sslMode),
+		},
+	}, nil
+}
+
+func (r *HarborDatabaseSpec) getSSLModeForNotary() harbormetav1.PostgresSSLMode {
+	switch r.SSLMode { //nolint:exhaustive
+	case harbormetav1.PostgresSSLModeAllow:
+		return harbormetav1.PostgresSSLModePrefer
+	default:
+		return r.SSLMode
+	}
+}
+
+type NotaryComponentSpec struct {
+	// +kubebuilder:validation:Optional
+	Server harbormetav1.ComponentSpec `json:"server"`
+
+	// +kubebuilder:validation:Optional
+	Signer harbormetav1.ComponentSpec `json:"signer"`
+}
+
+type ExternalRedisSpec struct {
+	// +kubebuilder:validation:Required
+	Address string `json:"address"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:ExclusiveMinimum=true
+	// +kubebuilder:default=6379
+	Port int32 `json:"port,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Pattern="[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*"
+	PasswordRef string `json:"passwordRef,omitempty"`
+}
+
+func (r *HarborComponentsSpec) RedisDSN(component harbormetav1.ComponentWithRedis) OpacifiedDSN {
+	return OpacifiedDSN{
+		DSN:         fmt.Sprintf("redis://%s:%d/%d", r.Redis.Address, r.Redis.Port, component.Index()),
+		PasswordRef: r.Redis.PasswordRef,
+	}
+}
+
+type CoreComponentSpec struct {
+	harbormetav1.ComponentSpec `json:",inline"`
+
+	// +kubebuilder:validation:Required
+	TokenIssuer cmmeta.ObjectReference `json:"tokenIssuer,omitempty"`
+}
+
+type JobServiceComponentSpec struct {
+	harbormetav1.ComponentSpec `json:",inline"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:default=10
+	WorkerCount int32 `json:"workerCount,omitempty"`
+}
+
+type RegistryComponentSpec struct {
+	harbormetav1.ComponentSpec `json:",inline"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=true
+	RelativeURLs *bool `json:"relativeURLs,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	StorageMiddlewares []RegistryMiddlewareSpec `json:"storageMiddlewares,omitempty"`
+}
+
+type ChartMuseumComponentSpec struct {
+	harbormetav1.ComponentSpec `json:",inline"`
+}
+
+type ClairComponentSpec struct {
+	harbormetav1.ComponentSpec `json:",inline"`
+
+	// +kubebuilder:validation:Optional
+	// One of clair redis dsn or global redis component must be specified
+	Redis *OpacifiedDSN `json:"redis,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Type="string"
+	// +kubebuilder:validation:Pattern="([0-9]+h)?([0-9]+m)?([0-9]+s)?([0-9]+ms)?([0-9]+us)?([0-9]+µs)?([0-9]+ns)?"
+	// +kubebuilder:default="12h"
+	Interval *metav1.Duration `json:"updatersInterval,omitempty"`
+}
+
+type TrivyComponentSpec struct {
+	harbormetav1.ComponentSpec `json:",inline"`
+}
+
+type HarborStorageImageChartStorageSpec struct {
```

Comment:
Simialr question to the storage divers supported list as chartmuseum component. Only 3 storage drivers selected.